io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly3pro/IoShelly3ProImpl.java

```diff
@@ -0,0 +1,185 @@
+import io.openems.edge.common.component.AbstractOpenemsComponent;
+import io.openems.edge.common.component.OpenemsComponent;
+import io.openems.edge.common.event.EdgeEventConstants;
+import io.openems.edge.io.api.DigitalOutput;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.utils.JsonUtils;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "IO.Shelly.3Pro", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE //
+)
+@EventTopics({ //
+		EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE //
+})
+
+public class IoShelly3ProImpl extends AbstractOpenemsComponent
+		implements IoShelly3Pro, DigitalOutput, OpenemsComponent, EventHandler {
+
+	private final Logger log = LoggerFactory.getLogger(IoShelly3ProImpl.class);
+	private final BooleanWriteChannel[] digitalOutputChannels;
+	private String baseUrl;
+
+	@Reference
+	private BridgeHttpFactory httpBridgeFactory;
+	private BridgeHttp httpBridge;
+
+	public IoShelly3ProImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				DigitalOutput.ChannelId.values(), //
+				IoShelly3Pro.ChannelId.values() //
+		);
+		this.digitalOutputChannels = new BooleanWriteChannel[] { //
+				this.channel(IoShelly3Pro.ChannelId.RELAY_1), //
+				this.channel(IoShelly3Pro.ChannelId.RELAY_2), //
+				this.channel(IoShelly3Pro.ChannelId.RELAY_3), //
+		};
+	}
+
+	@Activate
+	protected void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.baseUrl = "http://" + config.ip();
+		this.httpBridge = this.httpBridgeFactory.get();
+
+		for (int i = 0; i < 3; i++) {
+			String url = this.baseUrl + "/rpc/Switch.GetStatus?id=" + i;
+			this.httpBridge.subscribeJsonEveryCycle(url, this::processHttpResult);
+		}
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		this.httpBridgeFactory.unget(this.httpBridge);
+		this.httpBridge = null;
+		super.deactivate();
+	}
+
+	@Override
+	public BooleanWriteChannel[] digitalOutputChannels() {
+		return this.digitalOutputChannels;
+	}
+
+	@Override
+	public String debugLog() {
+		var b = new StringBuilder();
+		var i = 1;
+		for (BooleanWriteChannel channel : this.digitalOutputChannels) {
+			String valueText;
+			var valueOpt = channel.value().asOptional();
+			if (valueOpt.isPresent()) {
+				valueText = valueOpt.get() ? "ON" : "OFF";
+			} else {
+				valueText = "Unknown";
+			}
+			b.append(valueText);
+			if (i < this.digitalOutputChannels.length) {
+				b.append("|");
+			}
+			i++;
+		}
+		return b.toString();
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE //
+			-> this.eventExecuteWrite();
+		}
+	}
+
+	private void processHttpResult(JsonElement result, Throwable error) {
```

Comment:
make sure properly handle everything inside this method if you use the method `HttpBridge.subscribeJsonEveryCycle(String,BiConsumer<JsonElement, Throwable>)` to handle only exactly one parameter is not null then the following test would fail
```
	@Test
	public void testProcessHttpResult() throws Exception {
		final var shelly = new IoShelly3ProImpl();
		new ComponentTest(shelly) //
				.addReference("httpBridge", new DummyBridgeHttp()) //
//		.addReference("httpBridgeFactory", new DummyBridgeHttpFactory()) //
				.activate(MyConfig.create() //
						.setId(COMPONENT_ID) //
						.setIp("127.0.0.1") //
						.build()) //
		;

		shelly.processHttpResult(null, new RuntimeException("Some http error"));
	}
```

you can also use the method `BridgeHttp.subscribeJsonEveryCycle(String, ThrowingConsumer<JsonElement, Exception>, Consumer<Throwable>)` then you always have a none null parameter and the exception would be properly handled. e. g.

```
this.httpBridge.subscribeJsonEveryCycle(url, result -> {
    // ... handle result
}, error -> {
    // ... handle error (also gets triggered if the result handle method throws an exception)
});
```

but thats up to you which version you want to use.